### PR TITLE
Support auto pushdown

### PR DESCRIPTION
**Pull request by hwanghw**
Port Spark snowflake connector's autopushdown.

All unit tests and integration tests pass.

### CODE REVIEW

**Comment by hwanghw** [comment]
@jsleight can you or someone review the changes.  Thanks!

**Review by jsleight — COMMENTED** [review]
Thanks so much for the contribution and I apologize for taking so long to get to this.

I've only made it through about half of the files so far, but figured I'd submit the WIP review

File: src/it/resources/log4j.properties [hunk 1]
```diff
@@ -1 +1,9 @@
-log4j.rootLogger=OFF
+# Set everything to be logged to the console
```
[the remark this answers is not in the document]
Comment (hwanghw) [reply src/it/resources/log4j.properties:1]: @jsleight, thank you for your review and comments.

I checked the output.  Only RedshiftWriter logs the credentials at info level.  This is not only for integration test but also for production.  I will push changes to remove the credentials from RedshiftWriter log.

File: src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftRelation.scala [hunk 2]
```diff
@@ -188,19 +155,221 @@ private[redshift] case class RedshiftRelation(
...
+    if(filesToRead.isEmpty) reader.schema(resultSchema)
+
+    reader.load(filesToRead: _*)
+    .rdd
+    .map{row =>
+      val typeConvertedRow = row.toSeq.zipWithIndex.map {
+        case (f, i) =>
+          parquetDataTypeConvert(f, resultSchema.fields(i).dataType)
+      }
+      InternalRow(typeConvertedRow: _*)
```
[the remark this answers is not in the document]
Comment (hwanghw) [reply src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftRelation.scala:321]: There are 2 reasons:
1) Need to convert Parquet's Decimal data type to  correct Java data type based on result schema.  Parquet reader's conversion could be wrong, e.g.  a decimal is converted to float but result schema is double.
2) Timestamp and timezone confusion with Spark, Parquet and Redshift
      // For date or timestamp without time zone, Redshift unloads timestamps
      // to Parquet as if they were UTC even if they are intended to represent local times.
      // We need to convert 17:00 UTC to 17:00 Spark time zone.

File: src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftRelation.scala [hunk 3]
```diff
@@ -188,19 +155,221 @@ private[redshift] case class RedshiftRelation(
...
+    try {
+      jdbcWrapper.executeInterruptibly(conn.prepareStatement(unloadSql))
+      SqlToS3TempCache.setS3Path(statement.statementString, newTempDir)
+    } finally {
+      conn.close()
+    }
+    Some(newTempDir)
+  }
+
+  private def checkRedshiftAndS3OnSameRegion(creds: AWSCredentialsProvider): Unit = {
```
[the remark this answers is not in the document]
Comment (hwanghw) [reply src/main/scala/io/github/spark_redshift_community/spark/redshift/RedshiftRelation.scala:277]: will do the change

File: CHANGELOG [hunk 4]
```diff
@@ -1,5 +1,9 @@
 # spark-redshift Changelog
 
+## 5.10.0 (2022-06-06)
+
+- Support autopushdown as did in Spark Snowflake Connector
```
[the remark this answers is not in the document]
Comment (hwanghw) [reply CHANGELOG:5]: I'd recommend to keep below 2 for auto pushdown for this PR.  Will make them generic after battle tested later. 
- unload as parquet
- maintaining the hashmap cache of queries which have already been run

File: src/main/scala/io/github/spark_redshift_community/spark/redshift/pushdown/querygeneration/DateStatement.scala [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+            ConstantString("day, (0 - (") +
+              convertStatement(days, fields) + ") )," +
+              convertStatement(startDate, fields)
+          )
+
+      // AddMonths can't be pushdown to redshift because their functionality is different.
+      // For example,
+      // On spark 2.3/2.4, "2015-02-28" +1 month -> "2015-03-31"
+      // On spark 3.0 and Redshift,     "2015-02-28" +1 month -> "2015-03-28"
+      case AddMonths(_, _) => null
```
[the remark this answers is not in the document]
Comment (hwanghw) [reply src/main/scala/io/github/spark_redshift_community/spark/redshift/pushdown/querygeneration/DateStatement.scala:38]: The comments are misleading.

 For example,  SELECT ADD_MONTHS (CAST ( '2015-02-28' AS DATE ) , 1 )
 
On Redshift, the result is "2015-03-31"
On spark 3, the result is "2015-03-28"

I will correct the comments.

I'll submit the all changes in one commit after I get all comments from you.

File: src/it/resources/log4j.properties [hunk 6]
```diff
@@ -1 +1,9 @@
...
+log4j.rootCategory=WARN, console
+log4j.appender.console=org.apache.log4j.ConsoleAppender
+log4j.appender.console.target=System.err
+log4j.appender.console.layout=org.apache.log4j.PatternLayout
+log4j.appender.console.layout.ConversionPattern=%d{yy/MM/dd HH:mm:ss} %p %c{1}: %m%n
+
+
+log4j.logger.io.github.spark_redshift_community.spark.redshift=INFO
```
Comment (Brooke-white) [inline src/it/resources/log4j.properties:9]: logging should be disable by default, as it was before

Comment (hwanghw) [reply src/it/resources/log4j.properties]: Done. It is now OFF by default.

**Pull request closed without merging**